### PR TITLE
webhook based API-Coverage tool: type analysis

### PR DESCRIPTION
**Pull request by dushyanthsc**
This changeset adds the next iteration in the webhook based API-Coverage
tool by introducing type analysis. type analysis builds the actual resource
tree. Each node type is expected to implement NodeInterface method
typeAnalyzer(t reflect.Type). Inside this method each node creates child nodes
based on its type, for e.g. StructKindNode creates one child for each field
defined in the struct.

### CODE REVIEW

File: tools/webhook-apicoverage/resourcetree/README.md [hunk 1]
```diff
@@ -1,8 +1,11 @@
 # Resource Tree
 
-resourcetree package contains types and interfaces that define a resource tree(n-ary tree based representation of an API resource). Each resource tree is composed of nodes which have data encapsulated inside [nodeData](node.go) and operations that can be performed on each node in the interface [INode](node.go). Type of a node is logically defined by reflect.Kind. Each node type is expected to satisfy the [INode](node.go) interface.  
+resourcetree package contains types and interfaces that define a resource tree(n-ary tree based representation of an API resource). Each resource tree is composed of nodes which have data encapsulated inside [nodeData](node.go) and operations that can be performed on each node in the interface [INode](node.go). Type of a node is logically defined by reflect.Kind. Each node type is expected to satisfy the [NodeInterface](node.go) interface.
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/README.md]:
```suggestion
resourcetree package contains types and interfaces that define a resource tree(n-ary tree based representation of an API resource). Each resource tree is composed of nodes which have data encapsulated inside [nodeData](node.go) and operations that can be performed on each node in the interface [NodeInterface](node.go). Type of a node is logically defined by reflect.Kind. Each node type is expected to satisfy the [NodeInterface](node.go) interface.  
```

File: tools/webhook-apicoverage/resourcetree/README.md [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
 
-ConnectedNodes represent connections between nodes that are of same type(reflect.Type) and belong to same package but span across different trees or branches of same tree. An example of ConnectedNodes would be v1alpha1.Route.Spec.Traffic and v1alpha1.Route.Status.Traffic, both these Traffic fields are of type v1alpha1.TrafficTarget, but are present in different paths inside the resource tree. ConnectedNodes connects these two nodes, and a outlining of this type would present the coverage across the two branches and gives a unified view of what fields are covered.
\ No newline at end of file
+ConnectedNodes represent connections between nodes that are of same type(reflect.Type) and belong to same package but span across different trees or branches of same tree. An example of ConnectedNodes would be v1alpha1.Route.Spec.Traffic and v1alpha1.Route.Status.Traffic, both these Traffic fields are of type v1alpha1.TrafficTarget, but are present in different paths inside the resource tree. ConnectedNodes connects these two nodes, and a outlining of this type would present the coverage across the two branches and gives a unified view of what fields are covered.
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/README.md]:
```suggestion
ConnectedNodes represent connections between nodes that are of same type(reflect.Type) and belong to same package but span across different trees or branches of same tree. An example of ConnectedNodes would be v1alpha1.Route.Spec.Traffic and v1alpha1.Route.Status.Traffic, both these Traffic fields are of type v1alpha1.TrafficTarget, but are present in different paths inside the resource tree. ConnectedNodes connects these two nodes, and an outlining of this type would present the coverage across the two branches and give an unified view of what fields are covered.
```

File: tools/webhook-apicoverage/resourcetree/README.md [hunk 3]
```diff
@@ -1,8 +1,11 @@
...
+
+## Type Analysis
+A Resource tree is built using type analysis. Each node type is expected to implement NodeInterface method *typeAnalyzer(t reflect.Type)*. Inside this method each node creates child nodes  based on its type, for e.g. StructKindNode creates one child for each field defined in the struct. Type analysis are defined inside [typeanalyzer_tests](typeanalyzer_test.go)
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/README.md]:
```suggestion
A Resource tree is built using type analysis. Each node type is expected to implement NodeInterface method *typeAnalyzer(t reflect.Type)*. Inside this method each node creates child nodes based on its type, for e.g. StructKindNode creates one child for each field defined in the struct. Type analysis are defined inside [typeanalyzer_tests](typeanalyzer_test.go)  
```

File: tools/webhook-apicoverage/resourcetree/basictypekindnode.go [hunk 4]
```diff
@@ -33,33 +33,39 @@ func (b *BasicTypeKindNode) getData() nodeData {
...
+		b.possibleEnum = true
+	}
+}
+
+func (b *BasicTypeKindNode) string(v reflect.Value) string {
 	switch v.Kind() {
 	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
 		if v.Int() != 0 {
-			return false, strconv.Itoa(int(v.Int()))
+			return strconv.Itoa(int(v.Int()))
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/basictypekindnode.go]: can we use a v.String() instead? https://golang.org/pkg/reflect/#Value.String

File: tools/webhook-apicoverage/resourcetree/node.go [hunk 5]
```diff
@@ -20,12 +20,13 @@ import (
 	"reflect"
 )
 
-//node.go contains types and interfaces pertaining to nodes inside resource tree.
+// node.go contains types and interfaces pertaining to nodes inside resource tree.
 
-//INode interface defines methods that can be performed on each node in the resource tree.
-type INode interface {
+// NodeInterface interface defines methods that can be performed on each node in the resource tree.
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/node.go]:
```suggestion
// NodeInterface defines methods that can be performed on each node in the resource tree.
```

File: tools/webhook-apicoverage/resourcetree/structkindnode.go [hunk 6]
```diff
@@ -20,7 +20,10 @@ import (
 	"reflect"
 )
 
-//StructKindNode represents nodes in the resource tree of type reflect.Kind.Struct
+const v1TimeType = "v1.Time"
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/structkindnode.go]: create a const block instead.

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
+package resourcetree
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: missing documentation

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+const basicTypeName = "BasicType"
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: const block

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+		return errors.New(fmt.Sprint("BaseTypeNodeVerification: Expected 2 children got only : ", len(data.children)))
+	}
+
+	if value, ok := data.children["field1"]; ok {
+		n := value.getData()
+		if !n.leafNode || n.fieldType.Kind() != reflect.String || n.fieldType.PkgPath() != "" || len(n.children) != 0 {
+			return errors.New("BaseTypeNodeVerification: Invalid Child Node for field1")
+		}
+	} else {
+		return errors.New("BaseTypeNodeVerification: field1 Child Not found")
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: what is field1? Should this be pointing to some node name for easier debugging?

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	child = child.getData().children["structPtr-ptr"]
+	if err := verifyBaseTypeNode(child.getData()); err != nil {
+		return fmt.Errorf("PtrNodeVerification: Invalid Ptr type found for field:structPtr-ptr - %v", err)
+	}
+
+	child = data.children["basePtr"]
+	if len(child.getData().children) != 1 {
+		return errors.New("PtrNodeVerification: Unexpected field type for field:basePtr")
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: add what type we got instead.

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	child = child.getData().children["basePtr-ptr"]
+	if child.getData().fieldType.Kind() != reflect.Float32 || !child.getData().leafNode || child.getData().fieldType.PkgPath() != "" || len(child.getData().children) != 0 {
+		return errors.New("PtrNodeVerification: Unexpected field type for field:basePtr-ptr")
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: add the typoe we got

Comment (srinivashegde86) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: to all the verifications :)

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	child := data.children["structArr"]
+	if child.getData().fieldType.Kind() != reflect.Slice || len(child.getData().children) != 1 {
+		return errors.New("ArrNodeVerification: Unexpected field type for field:structArr")
+	}
+
+	child = child.getData().children["structArr-arr"]
+	if err := verifyBaseTypeNode(child.getData()); err != nil {
+		return fmt.Errorf("ArrNodeVerification: Invalid ArrorSlice type found for field:structArr-arr - %v", err)
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: ArrorSlice? do you mean array or slice?

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+	child = child.getData().children["baseArr-arr"]
+	if child.getData().fieldType.Kind() != reflect.Bool || !child.getData().leafNode || child.getData().fieldType.PkgPath() != "" || len(child.getData().children) != 0 {
+		return errors.New("ArrNodeVerification: Unexpected field type for field:baseArr-arr")
+	}
+
+	return nil
+}
+
+func verifyOtherTypeNode(data nodeData) error {
+	if len(data.children) != 2 {
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: feels like we are doing the same thing in every verify method. Can we combine these into a single method and every verifyFoo calls that instead?

File: tools/webhook-apicoverage/resourcetree/timetypenode.go [hunk 14]
```diff
@@ -0,0 +1,21 @@
+package resourcetree
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/timetypenode.go]: documentation

File: tools/webhook-apicoverage/resourcetree/typeanalyzer_test.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
+package resourcetree
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/resourcetree/typeanalyzer_test.go]: doc

File: tools/webhook-apicoverage/resourcetree/arraykindnode.go [hunk 16]
```diff
@@ -30,7 +32,14 @@ func (a *ArrayKindNode ) getData() nodeData {
...
 
-func (a *ArrayKindNode ) initialize(field string, parent INode, t reflect.Type, rt *ResourceTree) {
+func (a *ArrayKindNode ) initialize(field string, parent NodeInterface, t reflect.Type, rt *ResourceTree) {
 	a.nodeData.initialize(field, parent, t, rt)
 	a.arrKind = t.Elem().Kind()
-}
\ No newline at end of file
+}
+
+func (a *ArrayKindNode) analyzeType(t reflect.Type) {
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/arraykindnode.go]: Changed it to buildChildNodes

File: tools/webhook-apicoverage/resourcetree/structkindnode.go [hunk 17]
```diff
@@ -29,6 +32,38 @@ func (s *StructKindNode) getData() nodeData {
 	return s.nodeData
 }
 
-func (s *StructKindNode) initialize(field string, parent INode, t reflect.Type, rt *ResourceTree) {
+func (s *StructKindNode) initialize(field string, parent NodeInterface, t reflect.Type, rt *ResourceTree) {
 	s.nodeData.initialize(field, parent, t, rt)
 }
+
+func (s *StructKindNode) analyzeType(t reflect.Type) {
+
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/structkindnode.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/structkindnode.go [hunk 18]
```diff
@@ -29,6 +32,38 @@ func (s *StructKindNode) getData() nodeData {
...
+			childNode.initialize(t.Field(i).Name, s, t.Field(i).Type, s.tree)
+		} else {
+			childNode = s.tree.createNode(t.Field(i).Name, s, t.Field(i).Type)
+		}
+		s.children[t.Field(i).Name] = childNode
+		childNode.analyzeType(t.Field(i).Type)
+	}
+}
+
+func (s *StructKindNode) analyzeTimeNode(t reflect.Type) bool {
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/structkindnode.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/resourcetree.go [hunk 19]
```diff
@@ -45,23 +46,20 @@ func (r *ResourceTree) createNode(field string, parent INode, t reflect.Type) IN
...
-	// For types that are part of the standard package, we treat them as leaf nodes and don't expand further.
-	//https://golang.org/pkg/reflect/#StructField.
-	if len(t.PkgPath()) == 0 {
-		nd.leafNode = true
-	}
+	return n
+}
 
-	return nd
+//BuildResourceTree builds a resource tree by calling into analyzeType method starting from root.
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/resourcetree.go]: Fixed

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+
+type combinedNodeType struct {
+	b baseType
+	a arrayType
+	p ptrType
+}
+
+func verifyBaseTypeNode(data nodeData) error {
+	if len(data.children) != 2 {
+		return errors.New(fmt.Sprint("BaseTypeNodeVerification: Expected 2 children got only : ", len(data.children)))
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Fixed

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	if value, ok := data.children["field1"]; ok {
+		n := value.getData()
+		if !n.leafNode || n.fieldType.Kind() != reflect.String || n.fieldType.PkgPath() != "" || len(n.children) != 0 {
+			return errors.New("BaseTypeNodeVerification: Invalid Child Node for field1")
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func verifyPtrNode(data nodeData) error {
+	if len(data.children) != 2 {
+		return fmt.Errorf("PtrNodeVerification: Expected 2 children got: %d", len(data.children))
+	}
+
+	child := data.children["structPtr"]
+	if len(child.getData().children) != 1 {
+		return errors.New("PtrNodeVerification: Unexpected field type for field:structPtr")
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Makes sense. Fixed.

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	child = child.getData().children["structPtr-ptr"]
+	if err := verifyBaseTypeNode(child.getData()); err != nil {
+		return fmt.Errorf("PtrNodeVerification: Invalid Ptr type found for field:structPtr-ptr - %v", err)
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/timetypenode.go [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+func (ti *TimeTypeNode) getData() nodeData {
+	return ti.nodeData
+}
+
+func (ti *TimeTypeNode) initialize(field string, parent NodeInterface, t reflect.Type, rt *ResourceTree) {
+	ti.nodeData.initialize(field, parent, t, rt)
+	ti.leafNode = true
+}
+
+func (ti *TimeTypeNode) analyzeType(t reflect.Type) {}
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/timetypenode.go]: TimeTypeNode is constructed after we the check we mae in analyzeTimeNode(isTimeNode).

File: tools/webhook-apicoverage/resourcetree/README.md (hunk 1 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/README.md]: Fixed.

File: tools/webhook-apicoverage/resourcetree/README.md (hunk 2 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/README.md]: Fixed.

File: tools/webhook-apicoverage/resourcetree/README.md (hunk 3 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/README.md]: Fixed.

File: tools/webhook-apicoverage/resourcetree/basictypekindnode.go (hunk 4 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/basictypekindnode.go]: v.String() returns "<bool Value>" (e.g. of bool but same for others) hence cannot be used.

File: tools/webhook-apicoverage/resourcetree/node.go (hunk 5 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/node.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/structkindnode.go (hunk 6 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/structkindnode.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/test_util.go (hunk 7 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Added.

File: tools/webhook-apicoverage/resourcetree/test_util.go (hunk 8 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Fixed.

File: tools/webhook-apicoverage/resourcetree/test_util.go (hunk 13 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/test_util.go]: Each type is different, the uber verification would have to take parameters specific to each test. Any addition of new verification logic would require changing that method and all other tests. whereever it is possible I have moved it to a common logic.

File: tools/webhook-apicoverage/resourcetree/timetypenode.go (hunk 14 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/timetypenode.go]: added.

File: tools/webhook-apicoverage/resourcetree/typeanalyzer_test.go (hunk 15 above)
Comment (dushyanthsc) [reply tools/webhook-apicoverage/resourcetree/typeanalyzer_test.go]: added.

File: tools/webhook-apicoverage/resourcetree/test_util.go [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+
+func verifyBaseTypeNode(logPrefix string, data nodeData) error {
+	if len(data.children) != 2 {
+		return fmt.Errorf("%s Expected 2 children got only : %d", logPrefix, len(data.children))
+	}
+
+	if value, ok := data.children["field1"]; ok {
+		n := value.getData()
+		if !n.leafNode || n.fieldType.Kind() != reflect.String || n.fieldType.PkgPath() != "" || len(n.children) != 0 {
+			return fmt.Errorf("%s Unexpected field:field1 : - %v of kind : %s", logPrefix, n, n.fieldType.Kind())
```
Comment (adrcunha) [inline tools/webhook-apicoverage/resourcetree/test_util.go]: This error message looks confusing. I suggest  clearly logging what's the expectation and what's the actual value. Here and everywhere.

**Comment by srinivashegde86** [comment]
/lgtm
/hold
LGTM but Adriano has a comment, so adding a hold. Do `/hold cancel` if you want to merge without changes.

**Pull request merged**